### PR TITLE
fix(typeahead): incoming data are not filtered after typeahead kicks-in

### LINKED ISSUE #3725: fix(typeahead): incoming data are not filtered after typeahead kicks-in
# Bug description:
The list is filtered according to the entered data using user's functionality in demo, not ngx-bootstrap component's functionality.

### Steps to reproduce:
- Open async demo for typeahead
- Type one of the states, which are in the list
- See demo code

# Actual result:

The list is filtered according to the entered data using user's functionality in demo
screenshot from 2018-02-08 13-01-57

# Expected result:

The list is filtered according to the entered data using ngx-bootstrap component's functionality

# Versions of Angular, and Bootstrap:
Angular: all supported
Bootstrap: all supported

### PR DESCRIPTION
**Pull request by IraErshova**
Fix filtering incoming data after typeahead kicks-in. Remove filtering functionality from demo code to ngx-bootstrap component's code.

Close #3725

# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated tests.
 - [x] added/updated API documentation.
 - [x] added/updated demos.

### CODE REVIEW

File: src/typeahead/typeahead.directive.ts [hunk 1]
```diff
@@ -344,7 +344,15 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
     this._subscriptions.push(
       this.keyUpEventEmitter
         .debounceTime(this.typeaheadWaitMs)
-        .mergeMap(() => this.typeahead)
+        .mergeMap((value: string) => {
+          const normalizedQuery = this.normalizeQuery(value);
+
+          return this.typeahead.map((data: any[]) => {
```
Comment (valorkin) [inline src/typeahead/typeahead.directive.ts]: `any`? - use types or interfaces
remove `return`s

File: src/typeahead/typeahead.directive.ts [hunk 2]
```diff
@@ -344,7 +344,15 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
     this._subscriptions.push(
       this.keyUpEventEmitter
         .debounceTime(this.typeaheadWaitMs)
-        .mergeMap(() => this.typeahead)
+        .mergeMap((value: string) => {
```
Comment (valorkin) [inline src/typeahead/typeahead.directive.ts]: `typeahead` can be almost anything, and you would knew about it, if we would have type on it :(
so this PR will break 99% of use cases

**Review by valorkin — CHANGES_REQUESTED** [review]

File: src/typeahead/typeahead.directive.ts (hunk 2 above)
Comment (IraErshova) [reply src/typeahead/typeahead.directive.ts]: `value: string` is not typing for `typeahead`. This code is correct and doesn't break anything

File: src/typeahead/typeahead.directive.ts (hunk 1 above)
Comment (IraErshova) [reply src/typeahead/typeahead.directive.ts]: `typeahead` has type any in the whole module. Adding type for this instance should be a separate issue.

**Review by daniloff200 — DISMISSED** [review]

**Comment by dmitry-zhemchugov** [comment]
Tested. Issues were fixed by @IraErshova . Ready to merge

**Review by daniloff200 — APPROVED** [review]

**Review by valorkin — APPROVED** [review]

**Comment by valorkin** [comment]
@dmitry-zhemchugov has to be tested with real XHR as a source of data

**Pull request merged**